### ISSUE TITLE
wasm plugin config update make envoy memory leak

### ISSUE THREAD

**Issue by LugiaChang** [envoyproxy/envoy#27336]
*Title*: *wasm plugin config update make envoy memory leak*

*Description*:
>I am updating the wasm configuration in envoyfilter and saving the configuration by parsing it in the on_configure phase of wasm. But when I pass a large configuration and constantly update it, I found that the memory of envoy keeps increasing and is not reclaimed.
![WX20230510-163631@2x](https://github.com/envoyproxy/envoy/assets/22168057/1a06192a-96e2-4254-b176-eb284bcd5d1d)

envoyfilter.yaml:
[wasm-old-envoyfilter.txt](https://github.com/envoyproxy/envoy/files/11447623/wasm-old-envoyfilter.txt)

rust wasm code about on_configure, and memory fluctuations and upward trends are more severe in the wasm written by tinygo:
```
impl RootContext for HttpHeadersRoot {
    fn get_type(&self) -> Option<ContextType> {
        Some(ContextType::HttpContext)
    }

    fn on_configure(&mut self, _: usize) -> bool {
        if let Some(config_bytes) = self.get_plugin_configuration() {
            if str::from_utf8(&config_bytes).is_err() {
                return false;
            }
            let t: PluginConfiguration =
                serde_json::from_str(str::from_utf8(&config_bytes).unwrap()).unwrap();
            self.configuration = Rc::new(t);
            return true;
        } else {
            return false;
        }
    }

    fn create_http_context(&self, _: u32) -> Option<Box<dyn HttpContext>> {
        Some(Box::new(HttpHeaders {
            configuration: Rc::clone(&self.configuration),
        }))
    }
}
```

*Repro steps*:
> Apply the envoyfilter yaml, then constantly update this plugin configuration to have it deployed to wasm.

*Versions*:
Any version.

**Comment by kyessenov** [comment]
Wasm is latched to requests - are you sure you don't have any draining / long requests for old configurations. 
Can you share wasm-related stats?

CC @PiotrSikora @mpwarres

**Comment by LugiaChang** [comment]
> 

I'm not very good at envoy, can you tell me how to get stats about wasm related data?

**Comment by kyessenov** [comment]
Query prometheus endpoint (:15000/stats/prometheus on istio) and search for strings with "wasm".

**Comment by LugiaChang** [comment]
> Query prometheus endpoint (:15000/stats/prometheus on istio) and search for strings with "wasm".

Probably these, is there anything missing?

# TYPE envoy_wasm_envoy_wasm_runtime_null_created counter
envoy_wasm_envoy_wasm_runtime_null_created{} 25

# TYPE envoy_wasm_envoy_wasm_runtime_v8_created counter
envoy_wasm_envoy_wasm_runtime_v8_created{} 1450

# TYPE envoy_wasm_remote_load_cache_hits counter
envoy_wasm_remote_load_cache_hits{} 0

# TYPE envoy_wasm_remote_load_cache_misses counter
envoy_wasm_remote_load_cache_misses{} 0

# TYPE envoy_wasm_remote_load_cache_negative_hits counter
envoy_wasm_remote_load_cache_negative_hits{} 0

# TYPE envoy_wasm_remote_load_fetch_failures counter
envoy_wasm_remote_load_fetch_failures{} 0

# TYPE envoy_wasm_remote_load_fetch_successes counter
envoy_wasm_remote_load_fetch_successes{} 0

# TYPE metric_cache_count counter
metric_cache_count{cache="hit",wasm_filter="stats_filter"} 79900
metric_cache_count{cache="miss",wasm_filter="stats_filter"} 16

# TYPE envoy_wasm_envoy_wasm_runtime_null_active gauge
envoy_wasm_envoy_wasm_runtime_null_active{} 28

# TYPE envoy_wasm_envoy_wasm_runtime_v8_active gauge
envoy_wasm_envoy_wasm_runtime_v8_active{} 40

# TYPE envoy_wasm_remote_load_cache_entries gauge
envoy_wasm_remote_load_cache_entries{} 0

**Comment by LugiaChang** [comment]
I made an attempt to dynamically obtain a large configuration through dispatch_http_call and found that the memory issue was significantly improved. However, it also led to other problems, such as this issue #27311.
![WX20230510-163612@2x](https://github.com/envoyproxy/envoy/assets/22168057/9bdbbccb-61e6-40ba-8779-d50d0727af85)

**Comment by LugiaChang** [comment]
@kyessenov Any clue here?

**Comment by PiotrSikora** [comment]
> Wasm is latched to requests - are you sure you don't have any draining / long requests for old configurations. Can you share wasm-related stats?

This could potentially be a leak of the configuration string and/or RootContext, which are not latched to a single request.

Perhaps this fixes it? https://github.com/proxy-wasm/proxy-wasm-cpp-host/pull/357

**Comment by LugiaChang** [comment]
> > Wasm is latched to requests - are you sure you don't have any draining / long requests for old configurations. Can you share wasm-related stats?
> 
> This could potentially be a leak of the configuration string and/or RootContext, which are not latched to a single request.
> 
> Perhaps this fixes it? [proxy-wasm/proxy-wasm-cpp-host#357](https://github.com/proxy-wasm/proxy-wasm-cpp-host/pull/357)

Does the rust sdk have this problem?

**Comment by LugiaChang** [comment]
> > Wasm is latched to requests - are you sure you don't have any draining / long requests for old configurations. Can you share wasm-related stats?
> 
> This could potentially be a leak of the configuration string and/or RootContext, which are not latched to a single request.
> 
> Perhaps this fixes it? [proxy-wasm/proxy-wasm-cpp-host#357](https://github.com/proxy-wasm/proxy-wasm-cpp-host/pull/357)

I am now using a new approach where I dynamically load the required configuration through the tick function inside the Wasm plugin, so the Wasm plugin config itself doesn't need to be updated every time by EnvoyFIlter. However, I noticed that the memory of the Istio-proxy still increases. I am puzzled, and I wonder if this PR can fix this issue?
